### PR TITLE
#1087 Allow interleaving of F# and F* sources in a project

### PR DESCRIPTION
**Pull request by mateuszbujalski**
Preserve the order of files for extraction and F# compilation

### CODE REVIEW

File: fsharp.extraction.targets [hunk 1]
```diff
@@ -17,70 +71,44 @@
...
+                - Compile item will be completely overriden
         -->
 
         <!-- Make sure the FSTAR_EXTRACTED_PATH directory exists before extraction -->
         <MakeDir Directories="$(FSTAR_EXTRACTED_PATH)" Condition="!Exists('$(FSTAR_EXTRACTED_PATH)')" />
-        <!-- <Exec Command="$(FSTAR_EXE) $(FSTAR_WARN_ERROR) $(FSTAR_EXTRACTION_FLAGS) $(FSTAR_FLAGS) @(FStarSourceFile, ' ')"></Exec> -->
-        <Exec Command="$(FSTAR_EXE) $(FSTAR_WARN_ERROR) $(FSTAR_EXTRACTION_FLAGS) $(FSTAR_FLAGS) %(FStarSourceFile.Identity) --extract &quot;+%(FStarSourceFile.Filename)&quot;"></Exec>
+        
+        <!-- Run extraction for .fst files -->
+        <!-- TODO: What about interfaces? Should we at least run verification for them? -->
```
Comment (mateuszbujalski) [inline fsharp.extraction.targets:89]: @nikswamy Is running the extraction for all fst files enough? Or should I do something with interface files as well? It seems to work fine, but it's still not entirely clear to me.

File: ulib/fs/FStar_Map.fs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+end))); domain = (FStar_Set.union m1.domain m2.domain)})
+
+(* TODO: Only implicit arguments at the start of a function are erased, whereas the others are extracted to unit and obj 
+         which makes extracted function unusable. See examples/hello/TestFSharp for a minimal example. 
+
+         Here, key should be a generic argument with a comparison constraint instead of obj/unit. 
+
+         A simple workaround would be to change the declaration of map_val in the FStar.Map.fsti so that 
+         '#key:eqtype' parameter is moved before any non-implicit parameters (i.e. before 'f').
+*)
```
Comment (mateuszbujalski) [inline ulib/fs/FStar_Map.fs:58]:
```
(* FStar.Map.fsti *)
val map_val: #val1:Type -> #val2:Type -> f:(val1 -> val2) -> #key:eqtype -> t key val1 -> Tot (t key val2)
```
@nikswamy Were you already looking at this particular problem? If not, I suppose it's a good candidate for me to look into next. It seems that all other modules are blocked by the issues around removing of unused parameters across modules.

**Comment by nikswamy** [comment]
Thanks @mateuszbujalski! This looks really great ... very cool that you an add a Compile target with .fst files in an .fsproj. I'm merging this into my working branch on this topic. Not sure I fully understand the issue on FStar.Map issue ... will look into it and reply.

**Comment by nikswamy** [comment]
Regarding the extraction of non-prefix quantified type variables, as in FStar.Map.map_val:

Yes, you're right. When F* extracts code to OCaml or F#, only prenex quantified type variables are extracted as type variables. The rest (since F# and OCaml do not directly support first-class polymorphism) are erased to `unit` and trigger the use of Prims.unsafe_coerce of Obj.magic.

For instance, we extract FStar.Map.map_val to

```
let map_val = (fun ( f  :  'uuuuuu16115  ->  'uuuuuu16116 ) ( key  :  unit ) ( m  :  (obj, 'uuuuuu16115) t ) -> {mappings = (FStar_FunctionalExtensionality.on_domain (fun ( x  :  obj ) -> (f (m.mappings x)))); domain = m.domain})
```

and a test client 

```
let test (f:int -> nat) (m:t int int) : t int nat = map_val f m
```

is extracted to

```

**Pull request merged**